File: backend/src/models/opportunities.py
"""Pydantic contracts for opportunities (gigs / podcast spots) and their journal.

The wire contract follows the project's Option-A rule: entities are referenced by id
(``organization_id``, ``talk_id``), catalog vocabularies by ``short_name``
(``opportunity_format``, ``comp_type``, ``current_status``, ``payment_status``, ``contact_role``).

``OpportunityInput`` — the shape for create and full-replace update (PUT) — is deliberately limited
to descriptive and money-*setup* fields. The lifecycle fields (``status``, ``payment_status``,
``paid_on``, and the derived ``closed_at``) never travel on it: they move only through the journaled
endpoints (``PATCH /{id}/status``, ``PATCH /{id}/payment``, ``POST /{id}/close``), which is what
keeps ``status_events`` a faithful one-row-per-move log and ``closed_at`` recomputed in exactly the
places that own it (DATABASE.md §4, DEV-PLAN slice 3 acceptance #1/#3/#5/#8).
"""

from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class OpportunityInput(BaseModel):
    """Writable descriptive/money-setup fields, for create and full-replace update.

    Parameters
    ----------
    title : str
        Display title for the gig; 1-255 characters.
    organization_id : int
        The host venue / podcast (entity FK — an opportunity always has one).
    opportunity_format : str
        ``opportunity_formats`` catalog short_name (workshop, podcast_spot, …).
    comp_type : str
        ``comp_types`` catalog short_name (paid, pro_bono, trade). On create it seeds the initial
        payment status (paid → unbilled; pro_bono/trade → n_a) server-side.
    talk_id : int or None
        Which talk was offered (entity FK, optional).
    event_date : date or None
        Scheduled date, if known.
    fee_amount : Decimal or None
        Agreed fee; non-negative, up to 10 digits with 2 decimals. None for pro bono / unknown.
    currency : str
        ISO 4217 code; 3 letters, defaults to USD.
    angle : str or None
        The pitch angle; seeded from the venue's ``how_to_approach`` at create, then editable.
    outcome : str or None
        Free-text outcome / result notes.
    """

    title: str = Field(min_length=1, max_length=255)
    organization_id: int
    opportunity_format: str = Field(min_length=1)
    comp_type: str = Field(min_length=1)
    talk_id: int | None = None
    event_date: date | None = None
    fee_amount: Decimal | None = Field(default=None, ge=0, max_digits=10, decimal_places=2)
    currency: str = Field(default="USD", min_length=3, max_length=3)
    angle: str | None = None
    outcome: str | None = None


class OpportunityCreateInput(OpportunityInput):
    """Create-only extension of :class:`OpportunityInput` carrying optional lifecycle *seeds*.

    ``OpportunityInput`` (reused by PUT) is deliberately lifecycle-free; create alone may seed where
    a gig starts, so a back-filled opportunity lands in the stage it is actually in — with its first
    ``status_events`` row recorded there (not a phantom ``researching`` → stage move) — and can name
    its lead contact in the same call. All three are optional; omitting them reproduces the previous
    behaviour (start in ``researching``, payment status derived from the comp type, no lead).

    Parameters
    ----------
    starting_status : str or None
        ``opportunity_statuses`` short_name to start in; must be a non-terminal board stage. None
        starts in ``researching``.
    payment_status : str or None
        ``payment_statuses`` short_name to start in. None derives it from ``comp_type``.
    lead_contact_id : int or None
        A contact to link as the lead on this gig (``is_primary``). None links no one.
    """

    starting_status: str | None = None
    payment_status: str | None = None
    lead_contact_id: int | None = None


class OpportunityStatusPatch(BaseModel):
    """Move an opportunity to a new board stage (``PATCH /{id}/status``).

    Journals exactly one ``status_events`` row per real move (a move to the current status is a
    no-op) and recomputes ``closed_at``. The target must be a board stage; cancelled / lost are
    reached through the Close flow, not here.

    Parameters
    ----------
    status : str
        Target ``opportunity_statuses`` short_name.
    """

    status: str = Field(min_length=1)


class OpportunityPaymentPatch(BaseModel):
    """Update payment state (``PATCH /{id}/payment``); recomputes ``closed_at``.

    Marking a delivered gig paid moves it to History (#4); correcting the payment back off a settled
    state clears ``closed_at`` and returns the card to the board (#5).

    Parameters
    ----------
    payment_status : str
        ``payment_statuses`` catalog short_name.
    paid_on : date or None
        Date the payment was received, if applicable.
    """

    payment_status: str = Field(min_length=1)
    paid_on: date | None = None


class OpportunityCloseInput(BaseModel):
    """Close an opportunity through the Close flow (``POST /{id}/close``).

    Writes a terminal ``status_events`` row **and** captures the reason as a note (#8).

    Parameters
    ----------
    status : str
        Terminal ``opportunity_statuses`` short_name — cancelled or lost.
    reason : str
        Why it closed; required, recorded on the status event and as a note.
    """

    status: str = Field(min_length=1)
    reason: str = Field(min_length=1)


class OpportunityContactInput(BaseModel):
    """Link a contact to an opportunity, with their role on this gig.

    ``is_primary`` here means "lead on this gig" — unrelated to
    ``contact_organizations.is_primary`` (the default contact for a venue).

    Parameters
    ----------
    contact_id : int
        The contact to link.
    contact_role : str or None
        ``contact_roles`` catalog short_name for this gig.
    is_primary : bool
        Whether this contact is the lead on this gig.
    """

    contact_id: int
    contact_role: str | None = None
    is_primary: bool = False


class OpportunityContactUpdate(BaseModel):
    """Update a linked contact's role on a gig (the contact is fixed by the path)."""

    contact_role: str | None = None
    is_primary: bool = False


class OpportunityNoteInput(BaseModel):
    """Add a dated note to an opportunity.

    Parameters
    ----------
    body : str
        The note text; required.
    occurred_at : datetime or None
        When it happened; defaults to now server-side when omitted.
    """

    body: str = Field(min_length=1)
    occurred_at: datetime | None = None


class OpportunityContact(BaseModel):
    """A contact linked to an opportunity, with their per-gig role, for detail responses."""

    contact_id: int
    name: str
    contact_role: str | None
    is_primary: bool


class OpportunityNote(BaseModel):
    """A dated note on an opportunity."""

    id: int
    body: str
    occurred_at: datetime
    created_at: datetime


class StatusEvent(BaseModel):
    """One entry in an opportunity's status journal (the lifecycle history)."""

    id: int
    status: str
    note: str | None
    occurred_at: datetime


class OpportunitySummary(BaseModel):
    """One card in the flat board payload (also a History row).

    The SPA buckets cards into columns by ``current_status`` (the flat-list board decision), and
    filters board vs History by ``closed_at``. Carries the money / payment chip fields.
    """

    id: int
    title: str
    organization_id: int
    organization_name: str
    organization_type: str
    talk_title: str | None
    opportunity_format: str
    current_status: str
    comp_type: str
    fee_amount: Decimal | None
    currency: str
    payment_status: str
    event_date: date | None
    paid_on: date | None
    closed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class Opportunity(OpportunityInput):
    """Full opportunity detail: descriptive fields plus lifecycle state and the journals."""

    id: int
    organization_name: str
    talk_title: str | None
    current_status: str
    payment_status: str
    paid_on: date | None
    closed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    contacts: list[OpportunityContact] = Field(default_factory=list)
    notes: list[OpportunityNote] = Field(default_factory=list)
    status_events: list[StatusEvent] = Field(default_factory=list)

File: frontend/src/api/opportunities.ts
import {
  useMutation,
  useQuery,
  useQueryClient,
  type QueryKey,
  type UseQueryResult,
} from '@tanstack/react-query';
import { useApi } from './client';
import { dashboardKeys } from './dashboard';
import type { FollowUpRiderInput } from './outreaches';

// Mirrors backend models/opportunities.py. Entities travel as ids (organization_id, talk_id),
// catalogs as short_names (opportunity_format, comp_type, current_status, payment_status). Money
// (fee_amount) is a Decimal string over the wire; dates/timestamps are ISO strings.

export interface OpportunityInput {
  title: string;
  organization_id: number;
  opportunity_format: string; // opportunity_formats catalog short_name
  comp_type: string; // comp_types catalog short_name
  talk_id?: number | null;
  event_date?: string | null;
  fee_amount?: string | null; // Decimal → string
  currency?: string;
  angle?: string | null;
  outcome?: string | null;
}

/** Create-only superset of OpportunityInput carrying optional lifecycle seeds (backend
 *  OpportunityCreateInput). Omitting each reproduces the default: start in researching, payment
 *  status derived from comp_type, no lead. Not accepted by PUT (update stays lifecycle-free). */
export interface OpportunityCreateInput extends OpportunityInput {
  starting_status?: string | null; // opportunity_statuses short_name (non-terminal)
  payment_status?: string | null; // payment_statuses short_name
  lead_contact_id?: number | null; // contact linked as lead (is_primary) on this gig
  /** Opt-in reminder about the new gig, created server-side **in the same transaction**. Omitted
   *  or null is the off state. Not a second request on purpose: two calls would leave the gig
   *  created and the reminder silently missing after the user asked for one. */
  follow_up?: FollowUpRiderInput | null;
}

export interface OpportunitySummary {
  id: number;
  title: string;
  organization_id: number;
  organization_name: string;
  organization_type: string; // organization_types short_name (for the card's venue chip)
  talk_title: string | null; // resolved talk name (null when no talk chosen)
  opportunity_format: string;
  current_status: string; // opportunity_statuses short_name — the SPA buckets by this
  /** Furthest stage ever reached (short_name). Not the current column: the dashboard funnel counts
   *  reached-or-beyond, so a gig cancelled after being booked still counts toward Booked while
   *  sitting in neither. Compare via the catalog's `sort_order`. Null only if a gig has no status
   *  events, and creation always writes one. */
  max_reached_status: string | null;
  comp_type: string;
  fee_amount: string | null;
  currency: string;
  payment_status: string;
  event_date: string | null;
  paid_on: string | null;
  closed_at: string | null; // null = active board, non-null = History
  created_at: string;
  updated_at: string;
}

export interface OpportunityContact {
  contact_id: number;
  name: string;
  contact_role: string | null; // contact_roles short_name — role on this gig
  is_primary: boolean; // lead on this gig
}

export interface OpportunityNote {
  id: number;
  body: string;
  occurred_at: string;
  created_at: string;
}

export interface StatusEvent {
  id: number;
  status: string; // opportunity_statuses short_name
  note: string | null; // close reason on terminal transitions
  occurred_at: string;
}

export interface Opportunity extends OpportunityInput {
  id: number;
  organization_name: string;
  talk_title: string | null;
  current_status: string;
  payment_status: string;
  paid_on: string | null;
  closed_at: string | null;
  created_at: string;
  updated_at: string;
  contacts: OpportunityContact[];
  notes: OpportunityNote[];
  status_events: StatusEvent[];
  responses: OpportunityResponseCount[];
}

/** One response counter on a gig — how many of that call to action it generated.
 *
 * Types with no stored row are **absent and mean zero**: the grid is rendered from the
 * `opportunity_response_types` catalog, not from this list, so a type nobody has used yet still
 * shows. Counters are not dated and are not a target — they track audience growth in aggregate,
 * and the per-response detail lives in legacy-tracker and GHL. */
export interface OpportunityResponseCount {
  response_type: string;
  count: number;
}

export interface OpportunityContactInput {
  contact_id: number;
  contact_role?: string | null;
  is_primary?: boolean;
}

export interface OpportunityContactUpdate {
  contact_role?: string | null;
  is_primary?: boolean;
}

export interface OpportunityNoteInput {
  body: string;
  occurred_at?: string | null;
}

/**
 * Server-side filters for the board/History list.
 *
 * `status` is where a gig sits now; `entered` is a status it has *ever* been through, optionally
 * bounded by `[enteredFrom, enteredTo)`. The second is what a Dashboard target tile opens, and it
 * has to be server-side: the summary payload carries no status-event dates, so the SPA could not
 * work it out from what it already has.
 */
export interface OpportunityFilters {
  closed?: boolean;
  status?: string;
  entered?: string;
  enteredFrom?: string;
  enteredTo?: string;
  /** Only gigs that produced at least one response — what the funnel's Responses row opens. A
   *  counter raised and then zeroed does not qualify, matching the number on that row. */
  hasResponses?: boolean;
}

const opportunityKeys = {
  all: ['opportunities'] as const,
  lists: () => ['opportunities', 'list'] as const,
  // The whole filter set is the cache key: two different windows are two different lists, and
  // keying on only part of it would serve one period's gigs under another's.
  list: (filters: OpportunityFilters) => ['opportunities', 'list', filters] as const,
  detail: (id: number) => ['opportunities', id] as const,
};

/** List opportunities as flat board / History cards. `closed` omitted returns both; `false` the
 *  active board (closed_at IS NULL); `true` History. See {@link OpportunityFilters}. */
export function useOpportunities(
  filters: OpportunityFilters = {},
): UseQueryResult<OpportunitySummary[]> {
  const api = useApi();
  const { closed, status, entered, enteredFrom, enteredTo, hasResponses } = filters;
  return useQuery({
    queryKey: opportunityKeys.list(filters),
    queryFn: async () => {
      const params = new URLSearchParams();
      if (closed !== undefined) params.set('closed', String(closed));
      if (status) params.set('status', status);
      if (entered) {
        params.set('entered', entered);
        // Only meaningful alongside `entered`, and the server ignores them without it.
        if (enteredFrom) params.set('entered_from', enteredFrom);
        if (enteredTo) params.set('entered_to', enteredTo);
      }
      if (hasResponses) params.set('has_responses', 'true');
      const qs = params.toString();
      return (
        await api<{ opportunities: OpportunitySummary[] }>(`/opportunities${qs ? `?${qs}` : ''}`)
      ).opportunities;
    },
  });
}

/** Load one opportunity's full detail (linked contacts, notes, status journal). */
export function useOpportunity(id: number): UseQueryResult<Opportunity> {
  const api = useApi();
  return useQuery({
    queryKey: opportunityKeys.detail(id),
    queryFn: () => api<Opportunity>(`/opportunities/${id}`),
  });
}

export function useCreateOpportunity() {
  const api = useApi();
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (data: OpportunityCreateInput) =>
      api<Opportunity>('/opportunities', { method: 'POST', body: JSON.stringify(data) }),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: opportunityKeys.lists() }),
  });
}

export function useUpdateOpportunity(id: number) {
  const api = useApi();
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (data: OpportunityInput) =>
      api<Opportunity>(`/opportunities/${id}`, { method: 'PUT', body: JSON.stringify(data) }),
    onSuccess: (updated) => {
      queryClient.setQueryData(opportunityKeys.detail(id), updated);
      queryClient.invalidateQueries({ queryKey: opportunityKeys.lists() });
    },
  });
}

export function useDeleteOpportunity() {
  const api = useApi();
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: number) =>
      api<{ deleted: boolean }>(`/opportunities/${id}`, { method: 'DELETE' }),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: opportunityKeys.lists() }),
  });
}

/**
 * Optimistic status move (the board drag). The card's ``current_status`` is flipped immediately in
 * every cached board/History list so it jumps columns without waiting for the server (acceptance
 * #1); a failed PATCH restores the snapshot, rolling the card back (acceptance #2). ``onSettled``
 * invalidates so a move that also *closes* the card (e.g. delivering an already-settled gig) drops
 * it off the board on refetch, even though the optimistic step only moved it.
 */
export function usePatchStatus() {
  const api = useApi();
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ id, status }: { id: number; status: string }) =>
      api<Opportunity>(`/opportunities/${id}/status`, {
        method: 'PATCH',
        body: JSON.stringify({ status }),
      }),
    onMutate: async ({ id, status }) => {
      await queryClient.cancelQueries({ queryKey: opportunityKeys.lists() });
      const snapshot = queryClient.getQueriesData<OpportunitySummary[]>({
        queryKey: opportunityKeys.lists(),
      });
      for (const [key, list] of snapshot) {
        if (!list) continue;
        queryClient.setQueryData<OpportunitySummary[]>(
          key,
          list.map((o) => (o.id === id ? { ...o, current_status: status } : o)),
        );
      }
      return { snapshot };
    },
    onError: (_err, _vars, ctx) => {
      ctx?.snapshot.forEach(([key, list]: [QueryKey, OpportunitySummary[] | undefined]) =>
        queryClient.setQueryData(key, list),
      );
    },
    onSuccess: (updated) => {
      queryClient.setQueryData(opportunityKeys.detail(updated.id), updated);
    },
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: opportunityKeys.all });
    },
  });
}

/** Update payment state (recomputes closed_at server-side); marking paid moves a delivered gig to
 *  History, correcting it back reopens it. Refreshes every list plus the detail. */
export function usePatchPayment() {
  const api = useApi();
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({
      id,
      payment_status,
      paid_on,
    }: {
      id: number;
      payment_status: string;
      paid_on?: string | null;
    }) =>
      api<Opportunity>(`/opportunities/${id}/payment`, {
        method: 'PATCH',
        body: JSON.stringify({ payment_status, paid_on }),
      }),
    onSuccess: (updated) => {
      queryClient.setQueryData(opportunityKeys.detail(updated.id), updated);
      queryClient.invalidateQueries({ queryKey: opportunityKeys.lists() });
    },
  });
}

/** Close an opportunity (cancelled / lost) with a required reason; moves it to History. */
export function useCloseOpportunity() {
  const api = useApi();
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ id, status, reason }: { id: number; status: string; reason: string }) =>
      api<Opportunity>(`/opportunities/${id}/close`, {
        method: 'POST',
        body: JSON.stringify({ status, reason }),
      }),
    onSuccess: (updated) => {
      queryClient.setQueryData(opportunityKeys.detail(updated.id), updated);
      queryClient.invalidateQueries({ queryKey: opportunityKeys.lists() });
    },
  });
}

// The linked-contact and note mutations return the updated opportunity detail and change only its
// nested data (not the board card), so they refresh the detail cache without invalidating lists.
//
// `invalidatesDashboard` is for the ones that are *not* purely local. A response counter feeds the
// funnel card's final row, so leaving the dashboard cache alone would show a stale count until
// something else happened to refetch it — the same gap that let the outreach tile read low before
// slice 11.
function useOpportunityDetailMutation<TVariables>(
  request: (api: ReturnType<typeof useApi>, variables: TVariables) => Promise<Opportunity>,
  options: { invalidatesDashboard?: boolean } = {},
) {
  const api = useApi();
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (variables: TVariables) => request(api, variables),
    onSuccess: (updated) => {
      queryClient.setQueryData(opportunityKeys.detail(updated.id), updated);
      if (options.invalidatesDashboard) {
        queryClient.invalidateQueries({ queryKey: dashboardKeys.all });
      }
    },
  });
}

export function useAddOpportunityContact() {
  return useOpportunityDetailMutation<{ oppId: number; data: OpportunityContactInput }>(
    (api, { oppId, data }) =>
      api<Opportunity>(`/opportunities/${oppId}/contacts`, {
        method: 'POST',
        body: JSON.stringify(data),
      }),
  );
}

export function useUpdateOpportunityContact() {
  return useOpportunityDetailMutation<{
    oppId: number;
    contactId: number;
    data: OpportunityContactUpdate;
  }>((api, { oppId, contactId, data }) =>
    api<Opportunity>(`/opportunities/${oppId}/contacts/${contactId}`, {
      method: 'PUT',
      body: JSON.stringify(data),
    }),
  );
}

export function useRemoveOpportunityContact() {
  return useOpportunityDetailMutation<{ oppId: number; contactId: number }>(
    (api, { oppId, contactId }) =>
      api<Opportunity>(`/opportunities/${oppId}/contacts/${contactId}`, { method: 'DELETE' }),
  );
}

export function useAddOpportunityNote() {
  return useOpportunityDetailMutation<{ oppId: number; data: OpportunityNoteInput }>(
    (api, { oppId, data }) =>
      api<Opportunity>(`/opportunities/${oppId}/notes`, {
        method: 'POST',
        body: JSON.stringify(data),
      }),
  );
}

export function useDeleteOpportunityNote() {
  return useOpportunityDetailMutation<{ oppId: number; noteId: number }>((api, { oppId, noteId }) =>
    api<Opportunity>(`/opportunities/${oppId}/notes/${noteId}`, { method: 'DELETE' }),
  );
}

/** Set one response counter to a value.
 *
 * Not a delta: a double-fired `+` lands on the same number rather than counting twice, which is
 * what makes the control safe to lean on — and why there is no separate delete, since lowering a
 * counter to zero is the removal. */
export function useSetResponseCount() {
  return useOpportunityDetailMutation<{ oppId: number; responseType: string; count: number }>(
    (api, { oppId, responseType, count }) =>
      api<Opportunity>(`/opportunities/${oppId}/responses/${responseType}`, {
        method: 'PUT',
        body: JSON.stringify({ count }),
      }),
    { invalidatesDashboard: true },
  );
}